File: src/lib/evaluate.ts
/**
 * Calls the Python evaluator via internal HTTP.
 *
 * Why internal HTTP: Vercel's Python runtime is a separate process from our
 * Node.js runtime. Same-project fetch is the supported cross-runtime IPC.
 * INTERNAL_EVAL_SECRET prevents anyone outside our project from hitting
 * /api/evaluate directly even though it's public-by-default on Vercel.
 */

import type { Audience, ContentType, Moment } from "./engine-taxonomy";
import { optionalEnv, requireEnv } from "./require-env";

export type EvaluateParams = {
  text: string;
  content_type?: ContentType;
  audience?: Audience;
  moment?: Moment;
  /** Calibration precedents the engine should inject as voice
   *  guidance in the LLM scan prompt. See src/lib/precedents.ts.
   *  Empty array (or omitted) → engine falls back to the universal
   *  voice rules from PR #252. Block 2c of the calibration plan. */
  precedents?: ReadonlyArray<{
    approved_text: string;
    sample_size: number;
  }>;
};

/**
 * One hop in the pipeline's reasoning chain. Mirrors the
 * Python-side `RationaleHop` dataclass in
 * `src/content_checker/models.py`. Added to the API envelope in
 * schema v1.2.0 (human-eval build plan Session 1); Session 21 first
 * typed it on the TS side.
 */
export type RationaleHop = {
  step: string;
  inputs: Record<string, unknown>;
  output: Record<string, unknown>;
  confidence: number | null;
  rule_versions: Record<string, string>;
  ambiguity_flag: string | null;
};

/** Three-state verdict (API v1.1.0+). */
export type Verdict = "pass" | "violation" | "review_recommended" | "error";

export type EvaluationResult = {
  content_type?: string;
  /** Legacy two-state; prefer `verdict`. */
  overall_verdict: "pass" | "fail" | "error";
  /** Three-state verdict (API v1.1.0+). */
  verdict?: Verdict;
  /** Typed subtype when verdict === "review_recommended" (API v1.1.0+). */
  review_reason?: string | null;
  violations: Array<{
    standard_id: string;
    rule: string;
    issue: string;
    suggestion: string;
    /** Per-violation severity (API v2.0.0+). One of "high"/"medium"/"low". */
    severity?: string;
    source?: string;
    /** Per-violation confidence (API v1.1.0+). */
    confidence?: number;
    /** Typed ambiguity signal (API v1.2.0+). */
    ambiguity_flag?: string | null;
    /** Snapshot of the rule text's version at scan time (API v1.2.0+). */
    rule_version?: string;
    /** Other standard IDs emitted together (API v1.2.0+). */
    related_standards?: string[];
    /** Canonical docs URL for the standard on docs.contentrx.io (API v1.7.0+). */
    docs_url?: string;
    /** Customer-facing category (API v2.5.0+). One of "Voice & tone",
     *  "Mechanics", "Structure", "Accessibility", "Inclusion",
     *  "Big picture". Derived in the engine from the substrate
     *  standard_id; team-rule additions don't carry it. */
    category?: string;
  }>;
  passes: Array<{ standard_id: string; rule: string }>;
  summary?: string;
  audience?: string;
  moment?: string;
  pipeline?: Record<string, number>;
  /** Ordered pipeline hops (API v1.2.0+). */
  rationale_chain?: RationaleHop[];
};

/** Token-cost telemetry from the engine response (audit M-24, PR 9).
 * `cache_creation_input` and `cache_read_input` report Anthropic
 * prompt-caching activity per call. Total billed input ≈
 * input + cache_creation_input + cache_read_input × 0.10.
 *
 * Pre-PR-9 engine versions don't emit the cache fields; the TS side
 * coerces them to 0 in /api/check. */
export type EngineTokens = {
  input: number;
  output: number;
  cache_creation_input?: number;
  cache_read_input?: number;
};

export type EvaluateResponse = {
  result: EvaluationResult;
  latency_ms: number;
  tokens: EngineTokens;
};

/**
 * Thrown when /api/evaluate returns a non-2xx status. The message is
 * intentionally narrow — just the prefix + status + statusText — so
 * no upstream response body (which can echo user text from a pydantic
 * ValidationError or similar) ends up in thrown-error messages, which
 * eventually land in `logSafeError` truncated at 200 chars. Operators
 * needing the body should consult Vercel function logs for the engine
 * side directly.
 */
export class UpstreamEvaluatorError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "UpstreamEvaluatorError";
    this.status = status;
  }
}

function internalEvaluateUrl(): string {
  // In production we REQUIRE an explicit INTERNAL_EVAL_URL. Falling back
  // to NEXT_PUBLIC_APP_URL lets a misconfigured hostname leak the
  // internal secret + user text to the wrong origin. Only the dev
  // localhost fallback is allowed. requireEnv treats "" as missing, so
  // a blank-but-set value fails loud here instead of silently leaking.
  if (process.env.NODE_ENV === "production") {
    const url = requireEnv("INTERNAL_EVAL_URL");
    return `${url.replace(/\/$/, "")}/api/evaluate`;
  }

  const base =
    optionalEnv("INTERNAL_EVAL_URL") ??
    optionalEnv("NEXT_PUBLIC_APP_URL") ??
    "http://localhost:3000";
  return `${base.replace(/\/$/, "")}/api/evaluate`;
}

export async function evaluate(
  params: EvaluateParams,
): Promise<EvaluateResponse> {
  const secret = requireEnv("INTERNAL_EVAL_SECRET");

  const res = await fetch(internalEvaluateUrl(), {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-internal-secret": secret,
    },
    body: JSON.stringify(params),
    cache: "no-store",
  });

  if (!res.ok) {
    throw new UpstreamEvaluatorError(
      `Evaluation failed: ${res.status} ${res.statusText}`,
      res.status,
    );
  }

  return (await res.json()) as EvaluateResponse;
}

export type ClassifyResponse = {
  result: { content_type: string; moment: string };
  latency_ms: number;
  tokens: EngineTokens;
};

/**
 * Classify-only call into the Python evaluator. Skips the full check
 * pipeline (no rule scan, no validation) and returns just the
 * (content_type, moment) pair. Used by the public /api/classify route
 * which the MCP server's `classify_moment` tool consumes — cheaper than
 * running a full evaluation purely to peek at the moment.
 */
export async function classify(text: string): Promise<ClassifyResponse> {
  const secret = requireEnv("INTERNAL_EVAL_SECRET");

  const res = await fetch(internalEvaluateUrl(), {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-internal-secret": secret,
    },
    body: JSON.stringify({ text, mode: "classify" }),
    cache: "no-store",
  });

  if (!res.ok) {
    throw new UpstreamEvaluatorError(
      `Classification failed: ${res.status} ${res.statusText}`,
      res.status,
    );
  }

  return (await res.json()) as ClassifyResponse;
}

export type SuggestFixParams = {
  text: string;
  // ADR 2026-04-25 — standard_id is now optional. Schema-2.0.0 client
  // surfaces (LSP, plugin, action, MCP) strip substrate before
  // forwarding; only server-side callers with substrate access can
  // supply one. The rewriter falls back to issue + current_suggestion
  // when absent. Caller must supply at least one of (standard_id,
  // issue, current_suggestion).
  standard_id?: string;
  rule?: string;
  issue?: string;
  current_suggestion?: string;
};

export type SuggestFixResponse = {
  result: { rewritten: string; standard_id: string | null };
  latency_ms: number;
  tokens: EngineTokens;
};

export type RewriteDocumentResponse = {
  result: {
    rewritten: string;
    // Schema 2.4.0: one-sentence diagnostic of the document's broad
    // weaknesses. Empty string when the LLM's JSON parse failed; the
    // rewrite still ships in that case.
    diagnostic: string;
  };
  latency_ms: number;
  tokens: EngineTokens;
};

/**
 * Document-tier holistic rewrite. Calls into the Python engine's
 * `rewrite_document` mode (see `src/content_checker/rewrite_document.py`),
 * which runs a single LLM call producing a clean version of the
 * input in the ContentRX house voice. The dashboard's Document tier
 * fires this in parallel with the regular check call so the
 * customer sees both findings AND a suggested rewrite.
 *
 * Schema 2.3.0. Failures are non-fatal: the route catches and
 * surfaces a null `suggested_rewrite` field with a warning, so the
 * regular check results still render even when the rewrite call
 * times out or hits a transient Anthropic error.
 */
export async function rewriteDocument(
  text: string,
): Promise<RewriteDocumentResponse> {
  const secret = requireEnv("INTERNAL_EVAL_SECRET");

  const res = await fetch(internalEvaluateUrl(), {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-internal-secret": secret,
    },
    body: JSON.stringify({ text, mode: "rewrite_document" }),
    cache: "no-store",
  });

  if (!res.ok) {
    throw new UpstreamEvaluatorError(
      `Rewrite-document failed: ${res.status} ${res.statusText}`,
      res.status,
    );
  }

  return (await res.json()) as RewriteDocumentResponse;
}

/**
 * Suggest-fix call into the Python evaluator. Rewrites a flagged
 * string to clear a specific standard's violation. BUILD_PLAN_v2
 * Session 17 — backs the public `/api/suggest-fix` route consumed
 * by the LSP code-action provider.
 */
export async function suggestFix(
  params: SuggestFixParams,
): Promise<SuggestFixResponse> {
  const secret = requireEnv("INTERNAL_EVAL_SECRET");

  const res = await fetch(internalEvaluateUrl(), {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-internal-secret": secret,
    },
    body: JSON.stringify({ ...params, mode: "suggest_fix" }),
    cache: "no-store",
  });

  if (!res.ok) {
    throw new UpstreamEvaluatorError(
      `Suggest-fix failed: ${res.status} ${res.statusText}`,
      res.status,
    );
  }

  return (await res.json()) as SuggestFixResponse;
}

File: api/evaluate.py
"""Vercel Python function — runs the content_checker pipeline.

Internal endpoint. Only callable with x-internal-secret header matching
INTERNAL_EVAL_SECRET. The TS /api/check route is the public hot path and
is responsible for auth, quota, rate limiting, and violation logging.
This function just evaluates.

POST /api/evaluate
Request:
  {
    "text": "Click here",
    "content_type": "button_cta"       # optional
    "audience": "product_ui"            # optional, default product_ui
    "moment": "decision_point"          # optional, auto-detected if omitted
  }

Response:
  {
    "result": { ...CheckResult.to_dict() },
    "latency_ms": 1234,
    "tokens": { "input": 500, "output": 200 }
  }
"""

from __future__ import annotations

import hmac
import json
import os
import sys
import traceback
from http.server import BaseHTTPRequestHandler

# content_checker is the Python engine package at the monorepo root
# (src/content_checker/). Vercel's project root is cwd (/var/task/) and
# __file__ is at /var/task/api/evaluate.py, so src/ is one directory up.
# `vercel.json` → functions.includeFiles bundles src/content_checker/**
# with this function so the import resolves at runtime.
_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
sys.path.insert(0, os.path.join(_ROOT, "src"))

# Module-level imports kept to the cheap path. Heavy imports (check,
# classify, load_standards, suggest_fix) move inside their handler
# branches so cold starts that only handle classify or suggest_fix
# don't pay for the LLM stack. Python's module cache avoids re-paying
# on warm instances.
from content_checker.api_utils import (  # noqa: E402
    PromptInjectionError,
    RateLimitedError,
    RequestTimeoutError,
)
from content_checker.moments import detect_moment  # noqa: E402


class handler(BaseHTTPRequestHandler):
    def do_POST(self) -> None:
        expected = os.environ.get("INTERNAL_EVAL_SECRET", "")
        if not expected:
            # Fail closed: without a shared secret this endpoint is a billing
            # DoS vector and a plaintext-leaking proxy. We log to stderr but
            # return a generic error to callers.
            print("INTERNAL_EVAL_SECRET is not set; refusing all requests", file=sys.stderr)
            return self._respond(500, {"error": "Server not configured"})

        provided = self.headers.get("x-internal-secret", "")
        # hmac.compare_digest is constant-time; provided != expected
        # short-circuits at the first mismatching byte and leaks length
        # and prefix information under repeated requests.
        if not hmac.compare_digest(provided.encode("utf-8"), expected.encode("utf-8")):
            return self._respond(401, {"error": "Unauthorized"})

        try:
            length = int(self.headers.get("content-length", "0"))
            raw = self.rfile.read(length).decode("utf-8") if length > 0 else ""
            body = json.loads(raw) if raw else {}
        except (ValueError, json.JSONDecodeError) as exc:
            return self._respond(400, {"error": f"Invalid JSON: {exc}"})

        mode = body.get("mode", "check")

        text = body.get("text")
        if not isinstance(text, str) or not text:
            return self._respond(400, {"error": "text is required"})

        # Suggest-fix mode: rewrite a flagged string to clear a specific
        # standard. BUILD_PLAN_v2 Session 17 — consumed by the LSP's
        # code-action provider and (eventually) the web dashboard's
        # rewrite-in-place action.
        if mode == "suggest_fix":
            # ADR 2026-04-25 — standard_id is now optional. Schema-2.0.0
            # client surfaces (LSP, plugin, action, MCP) strip substrate
            # and never carry it; the rewriter falls back to issue +
            # current_suggestion. The rewriter still needs SOMETHING to
            # anchor on, so refuse a request that supplies none of
            # (standard_id, issue, current_suggestion).
            standard_id_raw = body.get("standard_id")
            standard_id: str | None = (
                standard_id_raw
                if isinstance(standard_id_raw, str) and standard_id_raw
                else None
            )
            issue_raw = body.get("issue")
            current_suggestion_raw = body.get("current_suggestion")
            if not (
                standard_id
                or (isinstance(issue_raw, str) and issue_raw)
                or (
                    isinstance(current_suggestion_raw, str)
                    and current_suggestion_raw
                )
            ):
                return self._respond(
                    400,
                    {
                        "error": (
                            "At least one of standard_id, issue, or "
                            "current_suggestion is required for mode=suggest_fix"
                        )
                    },
                )
            # Lazy import — keeps catalog/classify cold starts fast.
            from content_checker.suggest_fix import suggest_fix  # noqa: PLC0415
            try:
                result = suggest_fix(
                    text=text,
                    standard_id=standard_id,
                    rule=body.get("rule"),
                    issue=body.get("issue"),
                    current_suggestion=body.get("current_suggestion"),
                )
            except PromptInjectionError as exc:
                return self._respond(400, {"error": str(exc)})
            except RateLimitedError as exc:
                return self._respond(503, {"error": str(exc)}, retry_after=30)
            except RequestTimeoutError as exc:
                return self._respond(504, {"error": str(exc)})
            except Exception:  # noqa: BLE001
                traceback.print_exc()
                return self._respond(500, {"error": "Suggestion failed"})

            return self._respond(
                200,
                {
                    "result": {
                        "rewritten": result.rewritten,
                        "standard_id": standard_id,
                    },
                    "latency_ms": result.latency_ms,
                    "tokens": {
                        "input": result.input_tokens,
                        "output": result.output_tokens,
                    },
                },
            )

        # Document rewrite mode: produces a holistic clean version of
        # the input in the ContentRX house voice. The dashboard's
        # Document tier calls this in parallel with the regular check
        # so the customer sees both findings AND a suggested rewrite.
        # See rewrite_document.py for the prompt + output contract.
        if mode == "rewrite_document":
            from content_checker.rewrite_document import (  # noqa: PLC0415
                rewrite_document,
            )
            try:
                result = rewrite_document(text=text)
            except PromptInjectionError as exc:
                return self._respond(400, {"error": str(exc)})
            except RateLimitedError as exc:
                return self._respond(503, {"error": str(exc)}, retry_after=30)
            except RequestTimeoutError as exc:
                return self._respond(504, {"error": str(exc)})
            except Exception:  # noqa: BLE001
                traceback.print_exc()
                return self._respond(500, {"error": "Rewrite failed"})

            return self._respond(
                200,
                {
                    "result": {
                        "rewritten": result.rewritten,
                        # Schema 2.4.0 — one-sentence diagnostic for the
                        # Document-tier verdict header. Empty string when
                        # the LLM's JSON output couldn't be parsed; the
                        # rewrite still ships in that case.
                        "diagnostic": result.diagnostic,
                    },
                    "latency_ms": result.latency_ms,
                    "tokens": {
                        "input": result.input_tokens,
                        "output": result.output_tokens,
                        "cache_creation_input": result.cache_creation_input_tokens,
                        "cache_read_input": result.cache_read_input_tokens,
                    },
                },
            )

        # Classify-only mode: cheap (~1 LLM call) helper used by the MCP
        # server's classify_moment tool. Skips the full check pipeline
        # so MCP clients can plan content without burning a quota slot
        # on every classification probe.
        if mode == "classify":
            # Lazy imports — catalog mode never pays for these.
            from content_checker.classify import classify  # noqa: PLC0415
            from content_checker.standards.loader import load_standards  # noqa: PLC0415
            try:
                content_types = load_standards().get("content_types", [])
                content_type, classify_latency_s, classify_tokens = classify(
                    text=text, content_types=content_types,
                )
                moment = detect_moment(text=text, content_type=content_type)
            except PromptInjectionError as exc:
                return self._respond(400, {"error": str(exc)})
            except RateLimitedError as exc:
                return self._respond(503, {"error": str(exc)}, retry_after=30)
            except RequestTimeoutError as exc:
                return self._respond(504, {"error": str(exc)})
            except Exception:  # noqa: BLE001
                traceback.print_exc()
                return self._respond(500, {"error": "Classification failed"})

            return self._respond(
                200,
                {
                    "result": {
                        "content_type": content_type,
                        "moment": moment,
                    },
                    "latency_ms": int(classify_latency_s * 1000),
                    "tokens": {
                        "input": int(getattr(classify_tokens, "input", 0)),
                        "output": int(getattr(classify_tokens, "output", 0)),
                        "cache_creation_input": int(getattr(classify_tokens, "cache_creation_input", 0)),
                        "cache_read_input": int(getattr(classify_tokens, "cache_read_input", 0)),
                    },
                },
            )

        # Lazy import — only check mode pays. Catalog/classify don't.
        from content_checker import check  # noqa: PLC0415
        # Block 2c (calibration plan): /api/check pre-fetches matching
        # precedents from suggestion_precedents and forwards them so
        # the engine can inject voice guidance into the LLM scan
        # prompt. Empty list (or omitted) → engine falls back to the
        # universal voice rules from PR #252. Schema:
        #   precedents: [{"approved_text": str, "sample_size": int}, ...]
        precedents = body.get("precedents") or []
        try:
            result, latency_s, tokens = check(
                text=text,
                content_type=body.get("content_type"),
                audience=body.get("audience", "product_ui"),
                moment=body.get("moment"),
                precedents=precedents,
            )
        except PromptInjectionError as exc:
            # Caller-side error: input contained our sentinel. Return 400
            # with the message so the caller can show it to the user
            # ("modify the input and retry"). Doesn't burn a quota slot.
            return self._respond(400, {"error": str(exc)})
        except RateLimitedError as exc:
            # Anthropic 429 after retries. /api/check should backoff,
            # not treat as a generic 500. Retry-After: 30s is a
            # reasonable starting point — Anthropic's headers carry the
            # real value but we don't surface them through the SDK.
            return self._respond(503, {"error": str(exc)}, retry_after=30)
        except RequestTimeoutError as exc:
            # Per-stage timeout exhausted. Lets /api/check distinguish
            # "engine slow" from "engine broken."
            return self._respond(504, {"error": str(exc)})
        except Exception:  # noqa: BLE001
            # Keep the full traceback in stderr (Vercel captures it,
            # Sentry ingests from there) but return a generic message
            # to the caller. The exception string can include file
            # paths, model names, Anthropic error bodies, or truncated
            # LLM output — none of which the TS caller should surface.
            # (ENG-H-01 from 2026-04-22 audit.)
            #
            # PR-193/PR-194 added internal-only `detail` + `traceback`
            # fields here as diagnostic aids during the prod 502
            # incident. Removed (audit P2): they're dead instrumentation
            # now that the engine is reliable, and skipping them keeps
            # the leak surface as small as possible. If we hit an
            # opaque error class again, reintroduce them as a 1-line
            # diagnostic hotfix per the
            # `feedback_diagnostic_detail_injection` memory pattern.
            traceback.print_exc()
            return self._respond(500, {"error": "Evaluation failed"})

        return self._respond(
            200,
            {
                "result": result.to_dict(),
                "latency_ms": int(latency_s * 1000),
                "tokens": {
                    "input": int(getattr(tokens, "input", 0)),
                    "output": int(getattr(tokens, "output", 0)),
                    "cache_creation_input": int(getattr(tokens, "cache_creation_input", 0)),
                    "cache_read_input": int(getattr(tokens, "cache_read_input", 0)),
                },
            },
        )

    # Silence default stderr logging from BaseHTTPRequestHandler —
    # Vercel captures our explicit logs; the default format is noisy.
    def log_message(self, format: str, *args: object) -> None:  # noqa: A002
        return

    def _respond(self, status: int, body: dict, *, retry_after: int | None = None) -> None:
        payload = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(payload)))
        if retry_after is not None:
            self.send_header("retry-after", str(retry_after))
        self.end_headers()
        self.wfile.write(payload)
